lib/nmatrix/shortcuts.rb

```diff
@@ -34,6 +34,92 @@
 
 class NMatrix
 
+  # Methods for generating magic matrix.
+  module MagicHelpers
+    class << self
+      def odd_magic(nm, shape)
+        row = shape - 1
+        col = shape / 2       
+        nm[row,col] = 1
+        (2..shape * shape).each do |index|
+          if nm[(row + 1) % shape,(col + 1) % shape] == 0
+            row = (row + 1) % shape
+            col = (col + 1) % shape
+          else
+            row = (row - 1 + shape) % shape
+          end
+            nm[row,col] = index
+        end
+      end
+    
+      def doubly_even_magic(nm, shape)
+        mini_square_num = shape / 4
+        count = 1     
+        inv_count = shape * shape
+        shape.times do |row|
+          shape.times do |col|
+            if col >= mini_square_num and col < shape - mini_square_num
+              if row >= mini_square_num and row < shape - mini_square_num
+      	        nm[row,col] = count
+              else 
+                nm[row,col] = inv_count
+              end
+            elsif row < mini_square_num or row >= shape - mini_square_num
+              nm[row,col] = count
+            else
+              nm[row,col] = inv_count
+            end
+            count += 1
+            inv_count -= 1  
+          end
+        end
+      end
+    
+      def singly_even_magic(nm, shape)
+        half_shape = shape / 2
+        complementary_pair = (shape - 2) / 4
+        swap_col = NMatrix.new([shape])
+        index = 0 
+        mini_magic = NMatrix.new([half_shape,half_shape], 0, dtype: nm.dtype)
```

Comment:
As mini_magic is passed in odd_magic method, I have assigned its dtype same as that of nm.